Another panic in CLI processIgnoreChanges

### What happened?

Seems very similar to https://github.com/pulumi/pulumi/issues/16403

```
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::e1::pulumi:pulumi:Stack::e1-dev]
panic: runtime error: index out of range [1] with length 1

goroutine 116 [running]:
github.com/pulumi/pulumi/sdk/v3/go/common/resource.PropertyPath.reset({0x14001ca5220, 0x1, 0x2}, {{0x106dc4180?, 0x1400251a510?}}, {{0x106dc4180?, 0x14000aa2fd8?}}, 0x0, 0x0)
        /home/runner/work/pulumi/pulumi/sdk/go/common/resource/properties_path.go:434 +0xa44
github.com/pulumi/pulumi/sdk/v3/go/common/resource.PropertyPath.reset({0x14001ca5210, 0x2, 0x3}, {{0x107283fa0?, 0x140009b7620?}}, {{0x107283fa0?, 0x14001cabd10?}}, 0x0, 0x0)
        /home/runner/work/pulumi/pulumi/sdk/go/common/resource/properties_path.go:532 +0x4b0
github.com/pulumi/pulumi/sdk/v3/go/common/resource.PropertyPath.reset({0x14001ca5200, 0x3, 0x4}, {{0x107283fa0?, 0x140009b7590?}}, {{0x107283fa0?, 0x14001cabdd0?}}, 0x0, 0x0)
        /home/runner/work/pulumi/pulumi/sdk/go/common/resource/properties_path.go:532 +0x4b0
github.com/pulumi/pulumi/sdk/v3/go/common/resource.PropertyPath.Reset(...)
        /home/runner/work/pulumi/pulumi/sdk/go/common/resource/properties_path.go:545
github.com/pulumi/pulumi/pkg/v3/resource/deploy.processIgnoreChanges(0x14001cabce0, 0x140009b7590, {0x1400220d150, 0x1, 0x14001607340?})
        /home/runner/work/pulumi/pulumi/pkg/resource/deploy/step_generator.go:1896 +0x18c
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*stepGenerator).generateSteps(0x14000eff7c0, {0x113800118, 0x1400220d720})
        /home/runner/work/pulumi/pulumi/pkg/resource/deploy/step_generator.go:568 +0x860
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*stepGenerator).GenerateSteps(0x14000eff7c0, {0x113800118?, 0x1400220d720?})
        /home/runner/work/pulumi/pulumi/pkg/resource/deploy/step_generator.go:281 +0x30
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*deploymentExecutor).handleSingleEvent(0x14000acc1e0, {0x10739b300, 0x1400220d720})
        /home/runner/work/pulumi/pulumi/pkg/resource/deploy/deployment_executor.go:465 +0x544
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*deploymentExecutor).Execute.func3(0x1400108f800, 0x14000acc1e0, 0x140004210a0, {0x1073c37f8, 0x14001eed180}, {0x1073c37c0, 0x140009cdb30})
        /home/runner/work/pulumi/pulumi/pkg/resource/deploy/deployment_executor.go:255 +0x178
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*deploymentExecutor).Execute(0x14000acc1e0, {0x1073c37c0, 0x140009cdb30})
        /home/runner/work/pulumi/pulumi/pkg/resource/deploy/deployment_executor.go:271 +0x300
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*Deployment).Execute(...)
        /home/runner/work/pulumi/pulumi/pkg/resource/deploy/deployment.go:587
github.com/pulumi/pulumi/pkg/v3/engine.(*deployment).run.func1()
        /home/runner/work/pulumi/pulumi/pkg/engine/deployment.go:337 +0x88
created by github.com/pulumi/pulumi/pkg/v3/engine.(*deployment).run in goroutine 1
        /home/runner/work/pulumi/pulumi/pkg/engine/deployment.go:336 +0x174
```



### Example


```typescript
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";

const cfg = new pulumi.Config();
const step = cfg.requireNumber("step");

console.log("step", step);

const current = aws.s3.getCanonicalUserId({});

const example = new aws.s3.BucketV2("example", {bucket: "my-tf-example-bucket-t0yv0-2024-fri-jul-2019"});

const exampleBucketOwnershipControls = new aws.s3.BucketOwnershipControls("example", {
    bucket: example.id,
    rule: {
        objectOwnership: "BucketOwnerPreferred",
    },
});

function concat<T>(a: T[], b: T[]): T[] {
    return a.concat(b);
}

const exampleBucketAclV2 = new aws.s3.BucketAclV2("example-bucket", {
    bucket: example.id,
    accessControlPolicy: {
        grants: concat(step > 0 ? [] : [
            {
                grantee: {
                    id: current.then(current => current.id),
                    type: "CanonicalUser",
                },
                permission: "READ",
            }
        ], [
            {
                grantee: {
                    type: "Group",
                    uri: "http://acs.amazonaws.com/groups/s3/LogDelivery",
                },
                permission: "READ_ACP",
            },
        ]),
        owner: {
            id: current.then(current => current.id),
        },
    },
}, {
    dependsOn: [exampleBucketOwnershipControls],
    ignoreChanges: step === 0 ? [] : [
        "accessControlPolicy.grants[*]"
    ]
});


```


```bash
#!/usr/bin/env bash

set -euo pipefail

pulumi destroy --yes
pulumi config set step 0
pulumi up --yes
pulumi config set step 1
pulumi up --yes

```

### Output of `pulumi about`

```
CLI          
Version      3.124.0
Go Version   go1.22.5
Go Compiler  gc

Plugins
KIND      NAME    VERSION
resource  aws     6.45.0
resource  awsx    2.13.0
resource  docker  4.5.4
resource  docker  3.6.1
language  nodejs  unknown

Host     
OS       darwin
Version  14.5
Arch     arm64

This project is written in nodejs: executable='/Users/anton/bin/node' version='v18.18.2'

Current Stack: anton-pulumi-corp/e1/dev

TYPE                                                    URN
pulumi:pulumi:Stack                                     urn:pulumi:dev::e1::pulumi:pulumi:Stack::e1-dev
pulumi:providers:aws                                    urn:pulumi:dev::e1::pulumi:providers:aws::default_6_45_0
aws:s3/bucketV2:BucketV2                                urn:pulumi:dev::e1::aws:s3/bucketV2:BucketV2::example
aws:s3/bucketOwnershipControls:BucketOwnershipControls  urn:pulumi:dev::e1::aws:s3/bucketOwnershipControls:BucketOwnershipControls::example
aws:s3/bucketAclV2:BucketAclV2                          urn:pulumi:dev::e1::aws:s3/bucketAclV2:BucketAclV2::example-bucket


Found no pending operations associated with dev

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/anton-pulumi-corp
User           anton-pulumi-corp
Organizations  anton-pulumi-corp, moolumi, demo, pulumi
Token type     personal

Dependencies:
NAME            VERSION
@pulumi/pulumi  3.125.0
@types/node     18.19.41
typescript      5.5.3
@pulumi/aws     6.45.0
@pulumi/awsx    2.13.0

Pulumi locates its logs in /var/folders/gd/3ncjb1lj5ljgk8xl5ssn_gvc0000gn/T/com.apple.shortcuts.mac-helper// by default
```

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/changelog/pending/20240722--engine--fix-panics-due-to-different-length-ignorechanges-arrays.yaml
+++ b/changelog/pending/20240722--engine--fix-panics-due-to-different-length-ignorechanges-arrays.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: engine
+  description: Fix panics due to different length `ignoreChanges` arrays

--- a/sdk/go/common/resource/properties_path.go
+++ b/sdk/go/common/resource/properties_path.go
@@ -424,14 +424,22 @@ func (p PropertyPath) reset(old, new PropertyValue, oldIsSecret, newIsSecret boo
 					return true
 				} else if new.IsArray() {
 					if old.IsArray() {
-						for i := range old.ArrayValue() {
-							v := old.ArrayValue()[i]
+						oldArray := old.ArrayValue()
+						newArray := new.ArrayValue()
+						// If arrays are of different length then this is a path failure because we can't
+						// synchronise the two values.
+						if len(oldArray) != len(newArray) {
+							return false
+						}
+
+						for i := range oldArray {
+							v := oldArray[i]
 							// If this was a secret value in old, but new isn't currently a secret context then we need
 							// to mark this reset value as secret.
 							if oldIsSecret && !newIsSecret {
 								v = MakeSecret(v)
 							}
-							new.ArrayValue()[i] = v
+							newArray[i] = v
 						}
 					}
 					return true

--- a/sdk/go/common/resource/properties_path_test.go
+++ b/sdk/go/common/resource/properties_path_test.go
@@ -161,25 +161,25 @@ func TestPropertyPath(t *testing.T) {
 			value := makeValue()
 
 			v, ok := parsed.Get(value)
-			assert.True(t, ok)
+			assert.True(t, ok, "Failed to get %v from %v", parsed, value)
 			assert.False(t, v.IsNull())
 
 			ok = parsed.Delete(value)
-			assert.True(t, ok)
+			assert.True(t, ok, "Failed to delete %v from %v", parsed, value)
 
 			ok = parsed.Set(value, v)
-			assert.True(t, ok)
+			assert.True(t, ok, "Failed to set %v in %v", v, parsed)
 
 			u, ok := parsed.Get(value)
-			assert.True(t, ok)
+			assert.True(t, ok, "Failed to get %v from %v", parsed, value)
 			assert.Equal(t, v, u)
 
 			vv := PropertyValue{}
 			vv, ok = parsed.Add(vv, v)
-			assert.True(t, ok)
+			assert.True(t, ok, "Failed to add %v at %v", v, parsed)
 
 			u, ok = parsed.Get(vv)
-			assert.True(t, ok)
+			assert.True(t, ok, "Failed to get %v from %v", parsed, vv)
 			assert.Equal(t, v, u)
 		})
 	}
@@ -724,7 +724,7 @@ func TestReset(t *testing.T) {
 			nil,
 		},
 		{
-			"Nested object wildcard reset fails",
+			"Nested object wildcard index reset fails",
 			PropertyPath{"root", "*", 0},
 			PropertyMap{"root": NewProperty(PropertyMap{
 				"passes": NewProperty(1.0),
@@ -737,7 +737,20 @@ func TestReset(t *testing.T) {
 			nil,
 		},
 		{
-			"Nested array wildcard reset fails",
+			"Nested array wildcard, new array is shorter fails",
+			PropertyPath{"root", "array", "*"},
+			PropertyMap{"root": NewProperty(PropertyMap{
+				"array": NewProperty([]PropertyValue{
+					NewProperty(1.0),
+				}),
+			})},
+			PropertyMap{"root": NewProperty(PropertyMap{
+				"array": NewProperty([]PropertyValue{}),
+			})},
+			nil,
+		},
+		{
+			"Nested array wildcard index reset fails",
 			PropertyPath{"root", "*", 0},
 			PropertyMap{"root": NewProperty([]PropertyValue{
 				NewProperty(1.0),
@@ -750,7 +763,7 @@ func TestReset(t *testing.T) {
 			nil,
 		},
 		{
-			"Nested wildcard, old array is longer fails",
+			"Nested array wildcard index, old array is longer fails",
 			PropertyPath{"root", "*", 0},
 			PropertyMap{"root": NewProperty([]PropertyValue{
 				NewProperty(1.0),
@@ -762,7 +775,7 @@ func TestReset(t *testing.T) {
 			nil,
 		},
 		{
-			"Nested wildcard, new array is longer fails",
+			"Nested array wildcard index, new array is longer fails",
 			PropertyPath{"root", "*", 0},
 			PropertyMap{"root": NewProperty([]PropertyValue{
 				NewProperty(1.0),

